include/kokkos-utils/callbacks/TimerListener.hpp

```diff
@@ -0,0 +1,209 @@
+#ifndef KOKKOS_UTILS_CALLBACKS__TIMERLISTENER_HPP
+#define KOKKOS_UTILS_CALLBACKS__TIMERLISTENER_HPP
+
+#include "Kokkos_NumericTraits.hpp"
+
+#include "kokkos-utils/callbacks/Listener.hpp"
+#include "kokkos-utils/callbacks/Matcher.hpp"
+#include "kokkos-utils/concepts/ExecutionSpace.hpp"
+#include "kokkos-utils/timer/Timer.hpp"
+
+namespace Kokkos::utils::callbacks
+{
+
+template <typename...>
+class TimerListener;
+
+/**
+ * @brief Listener that starts a timer when a begin event is received and stops it
+ *        when a corresponding end event is received.
+ *
+ * The start event is selected by requiring it to be of type @p BeginEventType and
+ * to be a match for the passed matcher. The stop event must be of type @p EndEventType
+ * and have the same ID as the begin event.
+ *
+ * The underlying timer @ref timer is a @ref Kokkos::utils::timer::Timer, instantiated
+ * for the execution space @p Exec. This class stores the execution space
+ * instance passed to the constructor, and uses it as the one passed to @ref timer.
+ *
+ * This class is well-suited for timing the execution of a parallel workload, such as a
+ * @c Kokkos parallel for, parallel reduce, parallel scan, or deep copy operation, to which
+ * an execution space instance is passed. In such a situation, the execution space instance
+ * passed to this class's constructor should be the same as the one passed to the workload.
+ */
+template <
+    Kokkos::utils::callbacks::BeginEvent BeginEventType, Kokkos::utils::callbacks::EndEvent EndEventType,
+    Kokkos::utils::callbacks::Matcher<Kokkos::Impl::type_list<BeginEventType>> MatcherType,
+    Kokkos::utils::concepts::ExecutionSpace Exec
+>
+class TimerListener<BeginEventType, EndEventType, MatcherType, Exec>
```

Comment:
I think we should be smarter than that.

Why not having 2 matchers, one for the start, one for the stop ?

Reply:
In our new design we should be able to support such use cases, because we now have a very general `BeginEndMatcherType` template argument. But our current specialisations `BeginEndTimerListener` and `RegionListener` still select based on a match for the "begin event" and then look for a "paired" "end event".